### PR TITLE
Refactor schedules mechanism 

### PR DESCRIPTION
**Pull request by Hedingber**
**The requirement:** support CRUD operations for schedules.

**Refactor the schedules mechanism:**
The exiting mechanism was very messy and could not be extended without almost completely re-writing it - so anyways a refactor was needed. as part of refactoring I did a little research on existing projects that may be useful.
The existing mechanism was very basic, it had an infinite loop running in a thread every 60 seconds, and checks whether any of the scheduled jobs should be run. it was using [croniter](https://pypi.org/project/croniter/) to calculate the next `datetime` to run out of the cron string given by the user.
Problems with the existing mechanism:
* Wasn't designed to CRUD at all
* doesn't support async functions, and doesn't integrate nicely with the our event loop
* croniter bug - cron format is not the official format -  https://github.com/taichino/croniter/issues/76

I found [APScheduler](https://pypi.org/project/APScheduler/), full description why I chose to use it in the end
TL;DR - it's giving a lot of the functionality out of the box, and since the alternative is to write it by my own, I prefer not to

**Note** - this PR breaks BC for the schedules API, so basically we should bump minor when releasing this, practically it wasn't really widely used yet, to be discussed 

**User POV**:
The user can still use the good old way to schedule function:
```
train_run = trainer.run(train_task, schedule='*/1 * * * *')
```
If he want to use the extended capabilities he'll need to use the cron trigger class:
```
from mlrun.api import schemas
cron_trigger = schemas.ScheduleCronTrigger(minute='*/1') 
train_run = trainer.run(train_task, schedule=cron_trigger)
```
To do CRUD operations:
```
from mlrun import get_run_db
get_run_db().get_schedules('default')
get_run_db().get_schedule('default', 'my-trainer')
get_run_db().delete_schedule('default', 'my-trainer')
```

**Other changes in this PR:**
* Commenting out CI dependencies installation cache - it has a bug - it don't "identify" when new requirements are added
* Changed from `pytest==5.3.1` to `pytest~=5.4` since it's needed to enable using `pytest-asyncio` which needed for the scheduler tests
* Moved the singletons from a single file (`mlrun/api/singletons.py`) to a several files under this package (`mlrun.api.utils.singletons`) in order to resolve circular imports error
* Changed the schedules table name from `schedules` to `schedules_v2` to prevent needing to take care of migrations - this means data loss of schedules on upgrade - this is ok since this feature wasn't really used
* Removed `mlrun/db/periodic.py` (and its tests and usage) this had nothing functional in it, looks like an "Hachana lemazgan" that is irrelevant anymore
* Removed irrelevant tests of the old schedules mechanism
* Moved `tests/clients/logger/test_logger.py` to `tests/utils/logger/test_logger.py` it was mistakenly under the clients package
* Moved `pydantic` to `requiurements.txt` since on of the schemas are now part of the `run` function
* Fixed several places where defaulting the default project was missing

**Choosing APScheduler:**
**Pros:**
* It gives the whole scheduled jobs mechanism - we don't need to maintain a loop and check the next time to run, we simply give it a function with cron trigger and it does everything for us
* Automatically handles async **and** regular functions (like FastAPI, if it's `async def` run on the event loop, `def` in an awaited threadpool)
* It offers a much more user friendly way to express cron trigger + extended options such as giving start/end date, giving timezone, giving jitter [details here](https://apscheduler.readthedocs.io/en/v3.6.3/modules/triggers/cron.html#module-apscheduler.triggers.cron)
* The project generally looks more mature

**Cons:**
* Its "control" operations doesn't yet fully support asyncio - https://github.com/agronholm/apscheduler/issues/360 + https://github.com/agronholm/apscheduler/issues/313 - since it only happens on api initialization there's no real worry of blocking the event loop
* since it's not using cron string, we will not be able (or it will be harder) to use things like https://github.com/Salamek/cron-descriptor

**Notes:**
* Ideally we would use its [sqlalchemy job store](https://github.com/agronholm/apscheduler/blob/v3.6.3/apscheduler/jobstores/sqlalchemy.py) so it would handle the storage for us as well, the problem is that using it requires that the function arguments will be serializable, and one of our arguments is `db_session`
* The APScheduler cron trigger does have a [from_crontab](https://github.com/agronholm/apscheduler/blob/v3.6.3/apscheduler/triggers/cron/__init__.py#L86) method but because of this bug I'm not sure whether to support it for now https://github.com/agronholm/apscheduler/issues/286 - **edit:** decided to support it, added

### CODE REVIEW

File: mlrun/api/api/endpoints/schedules.py [hunk 1]
```diff
@@ -1,16 +1,49 @@
...
+        schedule.name,
+        schedule.kind,
+        schedule.scheduled_object,
+        schedule.cron_trigger,
+    )
+    return Response(status_code=status.HTTP_201_CREATED)
+
+
+@router.get("/projects/{project}/schedules", response_model=schemas.Schedules)
+def get_schedules(
```
Comment (yaronha) [inline mlrun/api/api/endpoints/schedules.py:29]: @Hedingber i would use list_ for consistency

File: mlrun/api/db/sqldb/db.py [hunk 2]
```diff
@@ -350,13 +353,65 @@ def list_artifact_tags(self, session, project):
         )
         return [row[0] for row in query]
 
-    def store_schedule(self, session, data):
-        sched = Schedule()
-        sched.struct = data
-        self._upsert(session, sched)
+    def create_schedule(
+        self,
```
Comment (yaronha) [inline mlrun/api/db/sqldb/db.py:357]: @Hedingber what about `state` , i.e. allowing to create a schedule in a draft/standby state

File: mlrun/api/db/sqldb/models.py [hunk 3]
```diff
@@ -156,11 +159,30 @@ class Run(Base, HasStruct):
         start_time = Column(TIMESTAMP)
         labels = relationship(Label)
 
-    class Schedule(Base, HasStruct):
-        __tablename__ = "schedules"
+    class Schedule(Base):
+        __tablename__ = "schedules_v2"
```
Comment (yaronha) [inline mlrun/api/db/sqldb/models.py:163]: @Hedingber need to add state (active/standby/error/..)

File: mlrun/db/httpdb.py [hunk 4]
```diff
@@ -471,11 +502,15 @@ def remote_status(self, kind, selector):
 
         return resp.json()['data']
 
-    def submit_job(self, runspec, schedule=None):
+    def submit_job(
```
Comment (yaronha) [inline mlrun/db/httpdb.py:505]: @Hedingber how do we handle errors and repeated errors in scheduled tasks?

File: mlrun/api/utils/singletons/k8s.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from mlrun.k8s_utils import K8sHelper
+
+# TODO: something nicer
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/utils/singletons/k8s.py:3]: yeah, separate PR

[the remark this answers is not in the document]
Comment (omesser) [reply mlrun/api/utils/singletons/k8s.py:3]: ❤️ 
the best singleton is ~a dead~ no singleton
But, assuming we do end up with them, yeah, let's make them ☝️ 
> "different PR" - of course

File: mlrun/api/utils/scheduler.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        """
+        :return: a tuple (function, args, kwargs) to be used with the APScheduler.add_job
+        """
+
+        if scheduled_object_kind == schemas.ScheduledObjectKinds.job:
+            from mlrun.api.api.utils import submit
+
+            return submit, [db_session, scheduled_object], {}
+        if scheduled_object_kind == schemas.ScheduledObjectKinds.pipeline:
+            raise NotImplementedError("Pipeline scheduling Not implemented yet")
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/utils/scheduler.py:149]: removed the if

File: mlrun/db/httpdb.py [hunk 7]
```diff
@@ -399,6 +400,36 @@ def delete_runtime_object(
...
+        error_message = f'Failed creating schedule {project}/{schedule.name}'
+        self.api_call('POST', path, error_message, body=json.dumps(schedule.dict()))
+
+    def get_schedules(
+        self, project: str, kind: schemas.ScheduledObjectKinds = None
+    ) -> schemas.Schedules:
+        project = project or default_project
+        params = {'kind': kind}
+        path = f'projects/{project}/schedules'
+        error_message = f'Failed getting schedules {project} ? {kind}'
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/db/httpdb.py:416]: I hate this part as well, it's just me trying to get the information into the error message. I did `?` (and not `/`) so we can identify what's query param and what's path... 
now that we have a better logger, the `self.api_call` should except also `kwargs` for the error log, and then these will move there, will be done in another PR, currently leaving as is

File: mlrun/api/schemas/schedule.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
+from datetime import datetime
+from enum import Enum
+from typing import Optional, List, Dict, Union, Callable
+
+from pydantic import BaseModel
+
+
+class ScheduleCronTrigger(BaseModel):
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply mlrun/api/schemas/schedule.py:8]: Talked offline - keeping as is

File: mlrun/api/db/sqldb/db.py (hunk 2 above)
Comment (Hedingber) [reply mlrun/api/db/sqldb/db.py:357]: Talked offline, for now adding the `state` and `desired_state` field to the models and schemas, the state handling mechanism will be added in a followup PR

File: mlrun/api/db/sqldb/models.py (hunk 3 above)
Comment (Hedingber) [reply mlrun/api/db/sqldb/models.py:163]: same

File: mlrun/db/httpdb.py (hunk 4 above)
Comment (Hedingber) [reply mlrun/db/httpdb.py:505]: same
